Bump version of `kube-state-metrics` to `v2.0.0-rc.1`

L3n41c opened the pull request:
### What does this PR do?

Bump version of some dependencies. `kube-state-metrics`, `datadog-operator`, `watermarkpodautoscaler`…

### Motivation

We need to bump the version of `kube-state-metrics` for the new `kubernetes_state_core` check.

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

The features the most impacted by this dependencies upgrade and that deserve further validation are:
* `kubernetes_state_core` core check
* `DatadogMetric`

L3n41c committed:
Bump version of `kube-state-metrics` to `v2.0.0-rc.1`

L3n41c commented on go.mod:
```diff
@@ -162,71 +157,72 @@ require (
 replace github.com/vishvananda/netlink => github.com/DataDog/netlink v1.0.1-0.20210312173533-6d628a7fc6f3
+
+// Remove once the upstream PR is merged
+replace k8s.io/kube-state-metrics/v2 => github.com/L3n41c/kube-state-metrics/v2 v2.0.0-rc.1.0.20210409121934-c22976b826b2
```
TODO: Open a PR upstream and get rid of this reference to a personal fork.

L3n41c replied:
Here is the upstream PR : kubernetes/kube-state-metrics#1455

L3n41c commented on pkg/collector/corechecks/cluster/orchestrator/orchestrator_transformer.go:
```diff
@@ -91,8 +91,12 @@ func extractService(s *corev1.Service) *model.Service {
 
-	if s.Spec.IPFamily != nil {
-		message.Spec.IpFamily = string(*s.Spec.IPFamily)
+	if s.Spec.IPFamilies != nil {
+		strFamilies := make([]string, len(s.Spec.IPFamilies))
+		for i, fam := range s.Spec.IPFamilies {
+			strFamilies[i] = string(fam)
+		}
+		message.Spec.IpFamily = strings.Join(strFamilies, ", ")
```
In the new vendored version of `kubernetes`, the
```go
        // IPFamily is tombstoned to show why 15 is a reserved protobuf tag.
        // IPFamily *IPFamily `json:"ipFamily,omitempty" protobuf:"bytes,15,opt,name=ipFamily,Configcasttype=IPFamily"`
```
field has been replaced by:
```go
        IPFamilies []IPFamily `json:"ipFamilies,omitempty" protobuf:"bytes,19,opt,name=ipFamilies,casttype=IPFamily"`
```

I didn’t dare reflecting this structure change (`slice` instead of `pointer`) in `model.Service` as it might have an impact on the backend side.

Is concatenating the values like this OK ?

mfpierre replied:
concatenating the values should be OK on our side 👍 

@L3n41c do you know how the backward compatibility is handled with previous kube versions?
Should we still check for the old field if the new field is empty?

L3n41c replied:
The old field is just not available anymore: https://github.com/kubernetes/api/blob/v0.20.5/core/v1/types.go#L4210

clamoriniere committed:
Update compliance kubeapiserver unittest

clamoriniere committed:
Fix pkg/utils/profiling unittest

Force the `gopkg.in/DataDog/dd-trace-go.v1` package dependency
to `v1.23.1` as requested in go.mod.

vboulineau committed:
Fixes for DDM package

L3n41c committed:
Fix `pkg/compliance/checks/custom` unit tests

L3n41c committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into lenaic/upgrade_ksm_v2

L3n41c committed:
Fix `pkg/util/kubernetes/apiserver` unit tests

L3n41c committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into lenaic/upgrade_ksm_v2

L3n41c committed:
Fix LICENSE

L3n41c committed:
Fix `tests_ebpf`

L3n41c committed:
Update go dep replacement comment with the upstream PR reference

olivielpeau reviewed (approved):
Approving for Agent Core, only reviewed the dep updates and I don't see any that affects the scope of Agent Core.

We'll want the CI to pass before merging this of course :)

Nit, not a blocker: could you add `/pkg/util/containerd/` to the list of directories owned by container-integrations? https://github.com/DataDog/datadog-agent/blob/9b2a63ab8df2c741cd24c8bc114ce29cafd129a0/.github/CODEOWNERS#L158-L162

clamoriniere commented on pkg/clusteragent/externalmetrics/autoscaler_watcher_test.go:
```diff
@@ -22,8 +22,14 @@ import (
+	"github.com/DataDog/watermarkpodautoscaler/api/v1alpha1"
 )
 
+func init() {
+	autoscaler.AddToScheme(scheme)
+	v1alpha1.AddToScheme(scheme)
+}
+
```
since the `scheme` is define in "package" not a test file. maybe it is better to move the init also in the "package"

L3n41c replied:
In fact, this `scheme` is defined in a test file of the same package, in [`pkg/clusteragent/externalmetrics/datadogmetric_controller_test.go` here](https://github.com/DataDog/datadog-agent/pull/7857/files#diff-f5f5499a84c7aeb5c77b7fcad8a533bc8832a16c29f3bdfb7293eea7445e941cR32).

Inside `pkg/clusteragent/externalmetrics`, there are 2 test files:
* `autoscaler_watcher_test.go` and
* `datadogmetric_controller_test.go`

As they belong to the same package, there is only one `scheme` variable (which happens to be declared in `datadogmetric_controller_test.go`.
* `autoscaler_watcher_test.go` does the `AddToScheme(…)` for the HPA and the WPA and
* `datadogmetric_controller_test.go` does the `AddToScheme(…)` for the DDM.

In the end, there’s only one `scheme` variable for the tests only but I’ve scattered the `AddToScheme(…)` to the test file which is needing a given type. Does that make sense ?

clamoriniere commented on pkg/clusteragent/externalmetrics/utils.go:
```diff
@@ -96,3 +100,20 @@ func setQueryConfigValues(aggregator string, rollup int) {
 	queryConfigRollup = rollup
 }
+
+func StructureIntoDDM(obj interface{}, structDest *v1alpha1.DatadogMetric) error {
```
```suggestion
func UnstructuredIntoDDM(obj interface{}, structDest *v1alpha1.DatadogMetric) error {
```

clamoriniere replied:
maybe  you can also take directly an `*unstructured.Unstructured` as parameter

L3n41c replied:
:+1: Done: https://github.com/DataDog/datadog-agent/pull/7857/commits/067fca31ab65d678c7b2056c1a3438ed22b16113

clamoriniere commented on pkg/clusteragent/externalmetrics/utils.go:
```diff
@@ -96,3 +100,20 @@ func setQueryConfigValues(aggregator string, rollup int) {
+	return runtime.DefaultUnstructuredConverter.FromUnstructured(unstrObj.UnstructuredContent(), structDest)
+}
+
+func StructureFromDDM(structIn *v1alpha1.DatadogMetric, unstructOut *unstructured.Unstructured) error {
```
```suggestion
func UnstructuredFromDDM(structIn *v1alpha1.DatadogMetric, unstructOut *unstructured.Unstructured) error {
```

L3n41c replied:
:+1: Done: https://github.com/DataDog/datadog-agent/pull/7857/commits/067fca31ab65d678c7b2056c1a3438ed22b16113

clamoriniere commented on pkg/collector/corechecks/cluster/ksm/kubernetes_state.go:
```diff
@@ -28,8 +28,8 @@ import (
 	"k8s.io/client-go/tools/cache"
-	"k8s.io/kube-state-metrics/pkg/allowdenylist"
-	"k8s.io/kube-state-metrics/pkg/options"
+	"k8s.io/kube-state-metrics/v2/pkg/allowdenylist"
```
🥇

clamoriniere commented on pkg/compliance/checks/kubeapiserver_check_test.go:
```diff
@@ -14,11 +14,98 @@ import (
+	return nil
+}
+
+func init() {
+	schemeBuilder := runtime.NewSchemeBuilder(addKnownTypes)
+	schemeBuilder.AddToScheme(scheme)
+}
```
you could move this function on top of the file

L3n41c replied:
:+1: Done: https://github.com/DataDog/datadog-agent/pull/7857/commits/067fca31ab65d678c7b2056c1a3438ed22b16113

clamoriniere commented on pkg/util/kubernetes/apiserver/util.go:
```diff
@@ -52,3 +52,12 @@ func StructureIntoWPA(obj interface{}, structDest *v1alpha1.WatermarkPodAutoscal
 	return runtime.DefaultUnstructuredConverter.FromUnstructured(unstrObj.UnstructuredContent(), structDest)
 }
+
+func StructureFromWPA(structIn *v1alpha1.WatermarkPodAutoscaler, unstructOut *unstructured.Unstructured) error {
```
```suggestion
func UnstructuredFromWPA(structIn *v1alpha1.WatermarkPodAutoscaler, unstructOut *unstructured.Unstructured) error {
```

L3n41c replied:
:+1: Done: https://github.com/DataDog/datadog-agent/pull/7857/commits/067fca31ab65d678c7b2056c1a3438ed22b16113

clamoriniere reviewed (commented):
👏  for this contribution.
few same nits

L3n41c committed:
Address code review comments

L3n41c committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into lenaic/upgrade_ksm_v2

mfpierre reviewed (approved):
LGTM for our files, we'll add a QA card to pay attention to service IP changes for this release

L3n41c committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into lenaic/upgrade_ksm_v2
